pkg/controller/attach_detach.go

```diff
@@ -573,154 +559,160 @@ func reportError(azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment, state azd
+		}
+	}
+	return err
+}
 
-				// for now, we don't empty the meta annotatinos after migrating them to status annotation for safety.
-				// note that this will leave some remnant garbage entries in meta annotations
+func generateVolumeInfo(volumeHandle string, volumeContext map[string]string) volumeInfo {
+	var volumeInfo volumeInfo
+	volumeInfo.handle = volumeHandle
+	volumeInfo.attributes = volumeContext
+	volumeInfo.syncedAzVolumeAttachments = make(map[string]bool)
+	return volumeInfo
+}
 
-				for k, v := range desiredAzVolumeAttachment.Annotations {
-					azVolumeAttachment.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
-				}
+func (r *ReconcileAttachDetach) loadOrStoreVolumeInfo(ctx context.Context, volumeName, azVolumeAttachmentName string) (string, map[string]string, volumeInfo, bool, error) {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	var err error
+	var volumeHandle string
+	var volumeContext map[string]string
+	var vInfo = volumeInfo{}
+	hasSynced := false
+	if val, ok := r.volumeInfos.Load(volumeName); ok {
+		vInfo = val.(volumeInfo)
+		if synced, hasAzVolumeAttachment := vInfo.syncedAzVolumeAttachments[azVolumeAttachmentName]; hasAzVolumeAttachment && synced {
+			hasSynced = true
+			return volumeHandle, volumeContext, vInfo, hasSynced, nil
+		}
+		volumeHandle = vInfo.handle
+		volumeContext = vInfo.attributes
+	} else {
+		w.Logger().Infof("getting volumeHandle and volumeContext from PersistentVolume for volume (%s)", volumeName)
+		volumeHandle, volumeContext, err = r.getVolumeInfoFromPV(ctx, volumeName)
+		if err != nil {
+			w.Logger().Errorf(err, "failed to find volume info from PersistentVolume for volume (%s)", volumeName)
+			return "", nil, vInfo, hasSynced, err
+		}
+		vInfo = generateVolumeInfo(volumeHandle, volumeContext)
+		r.volumeInfos.Store(volumeName, vInfo)
+	}
+	return volumeHandle, volumeContext, vInfo, hasSynced, nil
+}
 
-				azVolumeAttachment, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).Update(ctx, azVolumeAttachment, metav1.UpdateOptions{})
-				if err != nil {
-					w.Logger().Errorf(err, "failed to update AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			} else {
-				statusUpdateRequired = false
-			}
+func (r *ReconcileAttachDetach) recreateAzVolumeAttachmentFromVolumeInfo(ctx context.Context, volumeAttachmentName, azVolumeAttachmentName, volumeName, nodeName string) error {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	volumeHandle, volumeContext, volumeInfo, hasSynced, err := r.loadOrStoreVolumeInfo(ctx, volumeName, azVolumeAttachmentName)
+	if hasSynced {
+		// this azVolumeAttachment has already been recreated
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	attached, err := r.isAttached(ctx, volumeHandle, nodeName)
+	if err != nil {
```

Comment:
If the err is `getDisk` throttled, current code will return err. No azVolumeAttachment will be created. Then we have retry logic which will pick up the failed cases. Disadvantage: if all retries get throttled, we may miss the `azVolumeAttachment`. The recovery process can't recover all `azVolumeAttachment`s. I'm not sure how bad will it affect our driver's job.

Another way:
If the err is `getDisk` throttled, we still create the `azVolumeAttachment`. In the `recoverAzVolumeAttachment()` we re-check `isAttached()`. If it's not attached, we detach and delete the `azVolumeAttachment`. If it's still throttled, we just run normal recover process. Disadvantage: we may end up with creating an extra `azVolumeAttachment`. Advantage: we will not miss any `azVolumeAttachment`.

@edreed @sunpa93 @hccheng72 Please feel free to provide your input here.